pkg/usagestats/reporter.go

```diff
@@ -125,20 +137,26 @@ func (r *Reporter) running(ctx context.Context) error {
 			// If the send is failing since a long time and the report is falling behind,
 			// we'll skip this one and try to send the next one.
 			if time.Since(nextReportAt) >= r.reportSendInterval {
-				nextReportAt = nextReport(r.reportSendInterval, seed.CreatedAt, time.Now())
+				updateNextReportAt()
 				level.Info(r.logger).Log("msg", "failed to send anonymous usage stats report for too long, skipping to next report", "next_report_at", nextReportAt.String())
 				continue
 			}
 
+			// We're going to send the report. If we already have it, then it means it's a retry after a failure,
+			// otherwise we have to generate a new one.
+			if nextReport == nil {
+				nextReport = buildReport(seed, nextReportAt, r.reportSendInterval)
```

Comment:
Why don't we replace the `nextReport = nil` statements by `nextReport = buildReport()`, so we can get rid of `nil` value handling again.

While writing this, I just realized that we also have `nextReportAt` in `nextReport`.
So actually, what we can simplify to this:

```go

nextReport := buildReport(seed, nextReport(r.reportSeedInterval, seed.CreatedAt, time.Now(), r.reportSendInterval)
for {
	select {
	case <-ticker.C:
		if time.Now().Before(nextReport.At) {
			continue
		}
		if time.Since(nextReport.Interval) >= r.reportSendInterval {
			nextReport = buildReport(...)
			logger.Log(...)
			continue
		}
		if err := sendReport(ctx, nextReport); err != nil {
			logger.Log(...)
			continue
		}
		nextReport = buildReport(...)
	// ...
}
```

Reply:
I don't think the suggestion is logically equivalent. We want to generate the report as latest as possible (not before the `for` loop execution) and once the `nextReportAt` timestamp is reached.

Reply:
Oh, I see, you're right, we'd be always sending an _oldish_ report.